src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemoryAddress.java

```diff
@@ -83,9 +87,35 @@ default MemoryAddress address() {
     long segmentOffset(MemorySegment segment);
 
     /**
-     * Returns the raw long value associated to this memory address.
-     * @return The raw long value associated to this memory address.
-     * @throws UnsupportedOperationException if this memory address is associated with an heap segment.
+     * Returns a new confined native memory segment with given size, and whose base address is this address; the returned segment has its own temporal
+     * bounds, and can therefore be closed. This method can be very useful when interacting with custom native memory sources (e.g. custom allocators),
+     * where an address to some underlying memory region is typically obtained from native code
+     * (often as a plain {@code long} value). The returned segment will feature all <a href="#access-modes">access modes</a>
+     * (see {@link MemorySegment#ALL_ACCESS}), and its confinement thread is the current thread (see {@link Thread#currentThread()}).
+     * <p>
+     * This method is <em>restricted</em>. Restricted methods are unsafe, and, if used incorrectly, their use might crash
+     * the JVM or, worse, silently result in memory corruption. Thus, clients should refrain from depending on
+     * restricted methods, and use safe and supported functionalities, where possible.
+     *
+     * @param bytesSize the desired size.
+     * @return a new confined native memory segment with given base address and size.
+     * @throws IllegalArgumentException if {@code bytesSize <= 0}.
+     * @throws UnsupportedOperationException if this address is an heap address.
+     * @throws IllegalAccessError if the runtime property {@code foreign.restricted} is not set to either
+     * {@code permit}, {@code warn} or {@code debug} (the default value is set to {@code deny}).
+     */
+    default MemorySegment asSegmentRestricted(long bytesSize) {
+        if (bytesSize <= 0) {
+            throw new IllegalArgumentException("Invalid size : " + bytesSize);
+        }
+        Utils.checkRestrictedAccess("MemoryAddress.ofSegmentRestricted");
```

Comment:
Suggest this be the first check, so it fails independently of the arguments when permission is not granted.

Might be worth stating that closing does not free up any memory or resources. We can refer to the use of `withCleanupAction` if such action is required.